### PR TITLE
Feature application operations

### PR DESCRIPTION
**Pull request by derekolk**

### CODE REVIEW

**Comment by gkuchta** [comment]
LGTM

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy [hunk 1]
```diff
@@ -69,4 +65,17 @@ class Keys {
...
+    "${OpenstackCloudProvider.ID}:${Namespace.APPLICATIONS}:${application.toLowerCase()}"
+  }
+
+  static String getServerGroupKey(String serverGroupName, String account, String region) {
+    Names names = Names.parseName(serverGroupName)
+    "${OpenstackCloudProvider.ID}:${Namespace.SERVER_GROUPS}:${names.cluster}:${account}:${region}:${names.group}"
+  }
+
+  static String getClusterKey(String account, String application, String clusterName) {
+    "${OpenstackCloudProvider.ID}:${Namespace.CLUSTERS}:${account}:${application}:${clusterName}"
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy:79]: Is region not needed in the cluster key?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/CacheResultBuilder.groovy [hunk 2]
```diff
@@ -0,0 +1,106 @@
+/**
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/CacheResultBuilder.groovy]: Minor nitpick, but don't need the `**` for the license.

Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/CacheResultBuilder.groovy]: Actually, maybe that isn't minor, does this get picked up as the Javadoc for the class?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackInstanceCachingAgent.groovy [hunk 3]
```diff
@@ -67,7 +54,7 @@ class OpenstackInstanceCachingAgent implements CachingAgent, AccountAware {
   CacheResult loadData(ProviderCache providerCache) {
     Map<String, MutableCacheData> cachedInstances = MutableCacheData.mutableCacheMap()
 
-    List<Server> servers = this.account.credentials.provider.getInstances(region)
+    List<Server> servers = getClientProvider().getInstances(region)
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackInstanceCachingAgent.groovy]:
```
clientProvider.getInstances(region)
```

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/CacheResultBuilder.groovy (hunk 2 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/CacheResultBuilder.groovy]: I will fix it.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy (hunk 1 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/Keys.groovy:79]: No ... I modeled after google and aws ... both don't provide it.

**Pull request closed without merging**